### PR TITLE
Fix jbig2dec not working on Windows due to exclusive-locked temporary files.

### PR DESCRIPTION
**Pull request by kraptor**
This patch workarounds a python bug that renders jbig2dec unusable on Windows platforms.

This is due to temporary files being locked in exclusive mode. See here for more details:
> https://bugs.python.org/issue14243

Example of error:
> subprocess.CalledProcessError: Command '['jbig2dec', '-e', '-o', 'E:\\Users\\kraptor\\AppData\\Local\\Temp\\tmpcxmi9rk9', 'E:\\Users\\kraptor\\AppData\\Local\\Temp\\tmpf200vgs_']' returned non-zero exit status 1.

### CODE REVIEW

File: src/pikepdf/jbig2.py [hunk 1]
```diff
@@ -14,21 +15,42 @@
 
 
 def extract_jbig2(im_obj: pikepdf.Object, globals_obj: pikepdf.Object = None) -> Image:
-    with NamedTemp() as imgfile, NamedTemp() as globalfile, NamedTemp() as outfile:
+    # On Windows platform, the NamedTemporaryFile context manager creates the
+    # files in exclusive mode. Therefore, we need to fill the data of those
+    # files, then instruct NamedTemporaryFile to not to delete them.
```
Comment (mara004) [inline src/pikepdf/jbig2.py:20]: excess `to`

Comment (kraptor) [reply src/pikepdf/jbig2.py:20]: I'm not a native speaker. Please review everything and I will update as requested.

Comment (mara004) [reply src/pikepdf/jbig2.py:20]: I'm not a native speaker either ;) Anyone makes typos and that's not a problem, but when I see one I usually leave a note.
The rest of the text sounds good to me, though I only skimmed it.

Comment (kraptor) [reply src/pikepdf/jbig2.py]: I updated the PR. Let me know when everything OK and I will squash the changes for you :)

**Comment by jbarlow83** [comment]
Would you mind changing this to use a single `tempfile.TemporaryDirectory` with fixed filenames? That should avoid both the Windows issues and the need to do manual cleanup.

**Comment by kraptor** [comment]
Sure! Let me finish another task and I will work on this, should be easy.

File: src/pikepdf/jbig2.py [hunk 2]
```diff
@@ -4,30 +4,38 @@
...
+            with open(global_fname) as global_fd:
+                global_fd.write(globals_obj.read_raw_bytes())
+            args.append(global_fname)
+
+        args.append(image_fname)
 
         run(args, stdout=DEVNULL, check=True)
-        im = Image.open(outfile)
-        im.load()  # Load pixel data into memory so file can be closed
+        im = Image.open(output_fname)
```
Comment (kraptor) [inline src/pikepdf/jbig2.py:37]: I'm not very used to PIL: as we not "open" the filename (not the file descriptor)... do we need to do a "load()" afterwards or can we skip it?

**Comment by kraptor** [comment]
> Would you mind changing this to use a single `tempfile.TemporaryDirectory` with fixed filenames? That should avoid both the Windows issues and the need to do manual cleanup.

Updated as per your request. I have a doubt about PIL.Image, if you can validate the behavior is what I understand, then we can skip a "load" there.

**Comment by jbarlow83** [comment]
This is great.

The explicit Image.load is necessary - it tells Pillow to read the file into memory (because we are about to delete the file).

**Comment by kraptor** [comment]
Let me squash the commits, to have a clean history...

**Comment by kraptor** [comment]
Done. I added back the comment in the `im.load()` to not lose the context of why it's needed.

Let me know if there is anything else left.

**Comment by thadguidry** [comment]
@kraptor How do you install `jbig2dec` on Windows 10? or must it built, or I have to use Docker Ubuntu?

**Comment by kraptor** [comment]
> @kraptor How do you install `jbig2dec` on Windows 10? or must it built, or I have to use Docker Ubuntu?

@thadguidry : You can download the conda package for your Windows version, then use the .exe:
* https://anaconda.org/conda-forge/jbig2dec/files

When you need something obscure that doesn't have binaries for Windows:
* Maybe there is a conda package (https://anaconda.org/conda-forge/repo)
* Or gohlke has a package already built for you (https://www.lfd.uci.edu/~gohlke/pythonlibs/)

Happy coding!

**Comment by thadguidry** [comment]
@kraptor thanks, I figured as much... righto, or I just MSYS2 and build it :-)

**Comment by kraptor** [comment]
@thadguidry Well, jbig2dec is a very small utility, very easy to build. When you need something else or obscure dependencies it's when building might be a nightmare. Therefor I let it to the experts, specially when it's work-related and you are in a hurry.

**Pull request merged**